src/main/java/apoc/algo/pagerank/PageRankArrayStorageParallelCypher.java

```diff
@@ -0,0 +1,324 @@
+import org.neo4j.graphdb.RelationshipType;
+import org.neo4j.graphdb.Result;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.kernel.api.DataWriteOperations;
+import org.neo4j.kernel.api.exceptions.InvalidTransactionTypeKernelException;
+import org.neo4j.kernel.api.exceptions.legacyindex.AutoIndexingKernelException;
+import org.neo4j.kernel.api.exceptions.schema.ConstraintValidationKernelException;
+import org.neo4j.kernel.api.exceptions.schema.IllegalTokenNameException;
+import org.neo4j.kernel.api.properties.DefinedProperty;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.kernel.impl.core.ThreadToStatementContextBridge;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.Arrays;
+import java.util.concurrent.Future;
+
+import static apoc.algo.pagerank.PageRankUtils.toFloat;
+import static apoc.algo.pagerank.PageRankUtils.toInt;
+
+public class PageRankArrayStorageParallelCypher implements PageRank
+{
+    public static final int ONE_MINUS_ALPHA_INT = toInt( ONE_MINUS_ALPHA );
+    public static final int WRITE_BATCH=10_100;
+    public final int BATCH_SIZE = 100_000;
+    private final GraphDatabaseAPI db;
+    private final ExecutorService pool;
+    private String nodeCypher;
+    private String relCypher;
+    private int nodeCount;
+    private int relCount;
+
+
+    /*
+        1. Memory usage right now:
+            6 arrays : size of nodes.
+            2 arrays : size of relationships.
+            Memory usage: 6*N + 2*M
+        2. TODO:
+        Run this in parallel using ExecutorService.
+        Sort and divide into (number of threads) chunks and process it in parallel.
+     */
+
+    int [] nodeMapping;
+    int [] sourceDegreeData;
+
+    // Weighted Degrees.
+    // Don't absolutely need following two. Can be calculated on the fly.
+    int [] sourceWeightData;
+    int [] sourceChunkStartingIndex;
+
+    // Storing relationships
+    int [] relationshipTarget;
+    int [] relationshipWeight;
+
+    // Output arrays.
+    int [] pageRanks;
+    int [] previousPageRanks;
+
+    public PageRankArrayStorageParallelCypher(
+            GraphDatabaseAPI db,
+            ExecutorService pool,
+            String nodeCypher,
+            String relCypher)
+    {
+        this.pool = pool;
+        this.db = db;
+        this.relCypher = relCypher;
+        this.nodeCypher = nodeCypher;
+        readDataIntoArray(this.relCypher, this.nodeCypher);
+    }
+
+    private int getNodeIndex(int node) {
+        int index = Arrays.binarySearch(nodeMapping, node);
+        return index;
+    }
+
+    private int getDegree(int node) {
+        int degree = 0;
+        int index = Arrays.binarySearch(nodeMapping, node);
+        if (index >= 0) {
+            degree = sourceDegreeData[index];
+        }
+        return degree;
+    }
+
+    private void setDegree(int node, int degree) {
```

Comment:
This should happen while iterating over the mappingarray so there should be no id-lookup needed.